lib/new_api_prototype.dart

```diff
@@ -0,0 +1,895 @@
+// ignore_for_file: always_specify_types, prefer_const_constructors, public_member_api_docs
+
+import 'localization/l10n.dart';
+
+typedef Validator<T extends Object?> = String? Function(T);
+
+// Emptiness check validator:
+Validator<T?> isRequired<T extends Object>(
+  Validator<T>? v, {
+  String? isRequiredMessage,
+}) {
```

Comment:
I guess that here have a situation hard to define: use a _positional_ or _named_ parameter.

Positional: Is easy to use when combine with other validator but when no combine, generate things like `req(null)`, what is contraintuitive.
Named: Can use directly `req()` when not combine with other validator, but with combine, always need to write the name of parameter. This cons will be annoing when combine multiple validators.

Reply:
# Possibilities
I also reasoned about this situation. Initially I used the placeholder (ph) but I think it was even less intuitive. Then, I came up with the req(null), making the next validator positional parameter nullable. But it is also not intuitive. Those are some possibilities that I can think:
1. **isRequired and isRequiredWith**: use a direct validator (in the form `String? Function(T input)` ) for the pure `isRequired` and have a validator maker (a function that returns a pure validator) called isRequired<something> which would the the current `req`.
- The validator maker would have the following signature: `Validator<T?> isRequired<something><T extends Object>(Validator<T> v, {String? isRequiredMessage,})`
- The **something** field may be: with, and, ...
2. **isRequired(null)**: use the nullable input validator.
3. **isRequired({v, isRequiredMessage})**: use the named validator
4. **isRequired([v, isRequiredMessage])**: used optional positional arguments for both v and isRequiredMessage, but making it less intuitive about the isRequiredMessage

Reply:
I like the 4th option (`isRequired([v, isRequiredMessage])`), because isRequiredMessage is totally optional and can be specify in dart doc easyly as there are only two parameters.

Some example on real world of this, is the assert on Dart `assert(conditional, message)`

Reply:
ok